### PR TITLE
refactor(core): Get rid of 'fullpath'

### PR DESCRIPTION
**Pull request by niheaven**
Get rid of `fullpath`, just get absolute path when initialing `$globaldir`, `$scoopdir` and `$cachedir`, other dirs are derived from these three root ones.

### CODE REVIEW

File: lib/core.ps1 [hunk 1]
```diff
@@ -352,8 +352,12 @@ function url_remote_filename($url) {
...
 
 function ensure($dir) { if(!(test-path $dir)) { mkdir $dir > $null }; resolve-path $dir }
-function fullpath($path) { # should be ~ rooted
-    $executionContext.sessionState.path.getUnresolvedProviderPathFromPSPath($path)
+function Get-AbsolutePath {
+    param (
+        [String]
+        $Path
+    ) # should be ~ rooted
+    $ExecutionContext.SessionState.Path.GetUnresolvedProviderPathFromPSPath($Path)
```
Comment (Ash258) [inline lib/core.ps1]:
```suggestion
    return $ExecutionContext.SessionState.Path.GetUnresolvedProviderPathFromPSPath($Path)
```

**Comment by niheaven** [comment]
Removed `is_in_dir` test case will never be met when using `scoop`.

File: lib/core.ps1 [hunk 2]
```diff
@@ -352,8 +352,12 @@ function url_remote_filename($url) {
 }
 
 function ensure($dir) { if(!(test-path $dir)) { mkdir $dir > $null }; resolve-path $dir }
-function fullpath($path) { # should be ~ rooted
-    $executionContext.sessionState.path.getUnresolvedProviderPathFromPSPath($path)
+function Get-AbsolutePath {
+    param (
+        [String]
```
Comment (Ash258) [inline lib/core.ps1]: Use mandatory parameter

File: lib/core.ps1 [hunk 3]
```diff
@@ -352,8 +352,17 @@ function url_remote_filename($url) {
...
+    param (
+        [Parameter(Mandatory = $true)]
+        [String]
+        $Path
+    ) # should be ~ rooted
+    return $ExecutionContext.SessionState.Path.GetUnresolvedProviderPathFromPSPath($Path)
+}
+function fullpath($path) {
+    Show-DeprecatedWarning $MyInvocation 'Get-AbsolutePath'
+    Get-AbsolutePath -Path $path
```
Comment (Ash258) [inline lib/core.ps1]: It is not meant to be void

```suggestion
    return Get-AbsolutePath -Path $path
```

File: lib/core.ps1 [hunk 4]
```diff
@@ -352,8 +352,17 @@ function url_remote_filename($url) {
 }
 
 function ensure($dir) { if(!(test-path $dir)) { mkdir $dir > $null }; resolve-path $dir }
-function fullpath($path) { # should be ~ rooted
-    $executionContext.sessionState.path.getUnresolvedProviderPathFromPSPath($path)
+function Get-AbsolutePath {
+    param (
```
Comment (Ash258) [inline lib/core.ps1]: This is one parameter. It could be one liner.

File: lib/core.ps1 [hunk 5]
```diff
@@ -352,8 +352,17 @@ function url_remote_filename($url) {
 }
 
 function ensure($dir) { if(!(test-path $dir)) { mkdir $dir > $null }; resolve-path $dir }
-function fullpath($path) { # should be ~ rooted
-    $executionContext.sessionState.path.getUnresolvedProviderPathFromPSPath($path)
+function Get-AbsolutePath {
```
Comment (Ash258) [inline lib/core.ps1]: Add synopsis, Cmdletbinding, Outputtype, Pipeline support

File: lib/core.ps1 [hunk 6]
```diff
@@ -351,15 +351,32 @@ function url_remote_filename($url) {
...
+    [CmdletBinding()]
+    [OutputType([String])]
+    Param (
+        [Parameter(Mandatory = $true, ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true)]
         [String]
         $Path
-    ) # should be ~ rooted
-    return $ExecutionContext.SessionState.Path.GetUnresolvedProviderPathFromPSPath($Path)
+    )
+    Process {
```
Comment (Ash258) [inline lib/core.ps1]: What is advantage of having process for one line body function?

Comment (niheaven) [reply lib/core.ps1]: Just for pipeline. Indeed, there're only three usages in scoop now, and if not used in manifest's script, there is no need to support pipeline.

File: lib/core.ps1 [hunk 7]
```diff
@@ -351,9 +351,35 @@ function url_remote_filename($url) {
...
+            Get absolute path, even if not existed
+        .PARAMETER Path
+            Path to manipulate
+        .OUTPUTS
+            System.String
+                Absolute path, may or maynot existed
+    #>
+    [CmdletBinding()]
+    [OutputType([String])]
+    Param (
```
Comment (Ash258) [inline lib/core.ps1]:
```suggestion
    param (
```

File: lib/core.ps1 [hunk 8]
```diff
@@ -351,9 +351,35 @@ function url_remote_filename($url) {
...
+        [Parameter(Mandatory = $true, ValueFromPipeline = $true, ValueFromPipelineByPropertyName = $true)]
+        [String]
+        $Path
+    )
+    Process {
```
Comment (Ash258) [inline lib/core.ps1]:
```suggestion
    process {
```

File: lib/core.ps1 [hunk 9]
```diff
@@ -351,9 +351,35 @@ function url_remote_filename($url) {
...
 }
 
-function ensure($dir) { if(!(test-path $dir)) { mkdir $dir > $null }; resolve-path $dir }
-function fullpath($path) { # should be ~ rooted
-    $executionContext.sessionState.path.getUnresolvedProviderPathFromPSPath($path)
+function ensure($dir) { if (!(test-path $dir)) { mkdir $dir > $null }; resolve-path $dir }
+
+function Get-AbsolutePath {
+    <#
+        .SYNOPSIS
```
Comment (Ash258) [inline lib/core.ps1]: Synopsis is not indented from comment start in while codebase

https://github.com/lukesampson/scoop/search?q=synopsis&unscoped_q=synopsis

**Comment by Ash258** [comment]
This will break Excavator due to path glueing with hardcoded backslash.

image

image

**Pull request merged**